src/reactive/prometheus-apt-exporter.py

```diff
@@ -0,0 +1,208 @@
+PORT_NUMBER = "8089"
+
+
+@when("juju-info.connected")
+@when_not_all("apt-exporter.installed", "apt-exporter.started")
+def install_packages():
+    """Installs the prometheus-apt-exporter snap."""
+    hookenv.status_set("maintenance", "Installing software")
+    config = hookenv.config()
+    channel = config.get("snap_channel")
+    snap.install(SNAP_NAME, channel=channel, force_dangerous=False)
+    subprocess.check_call(["snap", "connect", "prometheus-apt-exporter:apt-exporter-files"])
+    hookenv.status_set("active", "Exporter installed and connected")
+    hookenv.open_port(PORT_NUMBER)
+    set_state("apt-exporter.installed")
+
+
+@hook("upgrade-charm")
+def upgrade():
+    """Reset the install state on upgrade, to ensure resource extraction."""
+    hookenv.status_set("maintenance", "Charm upgrade in progress")
+    remove_state("apt-exporter.installed")
+    remove_state("apt-exporter.started")
+    remove_state("apt-exporter.dashboard-registered")
+    update_dashboards_from_resource()
+    register_grafana_dashboards()
+
+
+@when_not("apt-exporter.started")
+@when_any("apt-exporter.installed", "config.changed")
+def start_snap():
+    """Configure snap.prometheus-apt-exporter.apt-exporter service."""
+    if not host.service_running(SVC_NAME):
+        hookenv.status_set("maintenance", "Service is down, starting")
+        hookenv.log("Service {} is down, starting...".format(SVC_NAME))
+        host.service_start(SVC_NAME)
+        hookenv.status_set("active", "Service started")
+        hookenv.log("start_snap() Service started")
+    else:
+        hookenv.status_set("active", "Ready")
+        set_state("apt-exporter.started")
+        hookenv.log("start_snap() apt-exporter.started")
+
+    update_dashboards_from_resource()
+
+    hookenv.log("Installed and set flag apt-exporter.started")
+
+
+@when("config.changed.snap_channel")
+def snap_channel_changed():
+    """Remove the state apt.exporter.installed if the snap channel changes."""
+    remove_state("apt-exporter.installed")
+    remove_state("apt-exporter.started")
+
+
+@when_all("apt-exporter.started", "scrape.available")
+def configure_scrape_relation(scrape_service):
+    """Connect prometheus to the the exporter for consumption."""
+    scrape_service.configure(PORT_NUMBER)
+    remove_state("apt-exporter.configured")
+
+
+@when("nrpe-external-master.changed")
+def nrpe_changed():
+    """Trigger nrpe update."""
+    remove_state("apt-exporter.configured")
+
+
+@when("apt-exporter.changed")
+def prometheus_changed():
+    """Trigger prometheus update."""
+    remove_state("apt-exporter.prometheus_relation_configured")
+    remove_state("apt-exporter.configured")
+
+
+@when("nrpe-external-master.available")
+@when_not("apt-exporter.configured")
+def update_nrpe_config(svc):
+    """Configure the nrpe check for the service."""
+    if not os.path.exists("/var/lib/nagios"):
+        hookenv.status_set("blocked", "Waiting for nrpe package installation")
```

Comment:
Setting unit status in reactive charms is tricky when it can come from different sources. For example in your case there are 3 functions that call `status_set` and you don't know in which order they'll be executed. For this reason, reactive charms ten to utilise flags and then final `atexit` decorated function that sets the final states after considering each possibility ([example](https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/blob/ce8cea2b9bc6d8dfd60a0d3455be3aa97fdd8eb9/reactive/kubernetes_control_plane.py#L778))